// victory-axis/lib/components/tick.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _lodash = require("lodash");

var _lodash2 = _interopRequireDefault(_lodash);

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _radium = require("radium");

var _radium2 = _interopRequireDefault(_radium);

var _victoryLabel = require("victory-label");

var _util = require("../util");

var orientationSign = {
  top: -1,
  left: -1,
  right: 1,
  bottom: 1
};

var anchorOrientation = {
  top: "end",
  left: "end",
  right: "start",
  bottom: "start"
};

var orientationVerticality = {
  top: false,
  bottom: false,
  left: true,
  right: true
};

var Tick = (function (_React$Component) {
  _inherits(Tick, _React$Component);

  function Tick() {
    _classCallCheck(this, _Tick);

    _get(Object.getPrototypeOf(_Tick.prototype), "constructor", this).apply(this, arguments);
  }

  _createClass(Tick, [{
    key: "getCalculatedValues",
    value: function getCalculatedValues(props) {
      var style = props.style.ticks;
      var tickSpacing = style.size + style.padding;
      var sign = orientationSign[props.orientation];
      var anchor = anchorOrientation[props.orientation];
      var isVertical = orientationVerticality[props.orientation];
      this.transform = isVertical ? "translate(0, " + props.position + ")" : "translate(" + props.position + ", 0)";
      this.x = isVertical ? sign * tickSpacing : 0;
      this.x2 = isVertical ? sign * style.size : 0;
      this.y = isVertical ? 0 : sign * tickSpacing;
      this.y2 = isVertical ? 0 : sign * style.size;
      this.textAnchor = isVertical ? anchor : "middle";
      this.verticalAnchor = isVertical ? "middle" : anchor;
    }
  }, {
    key: "evaluateStyle",
    value: function evaluateStyle(style) {
      var _this = this;

      return _lodash2["default"].transform(style, function (result, value, key) {
        result[key] = _lodash2["default"].isFunction(value) ? value.call(_this, _this.props.tick) : value;
      });
    }
  }, {
    key: "renderLabel",
    value: function renderLabel() {
      var isComponent = (0, _util.getRole)(this.props.label) === "label";
      var props = isComponent ? this.props.label.props : {};
      var style = props.style || this.props.style.tickLabels;
      var newProps = {
        x: this.x,
        y: this.y,
        textAnchor: props.textAnchor || this.textAnchor,
        verticalAnchor: props.verticalAnchor || this.verticalAnchor,
        style: this.evaluateStyle(style)
      };
      return isComponent ? _react2["default"].cloneElement(this.props.label, newProps) : _react2["default"].createElement(_victoryLabel.VictoryLabel, newProps, this.props.label);
    }
  }, {
    key: "renderTick",
    value: function renderTick(props) {
      return _react2["default"].createElement("line", {
        x: this.x,
        x2: this.x2,
        y: this.y,
        y2: this.y2,
        style: this.evaluateStyle(props.style.ticks)
      });
    }
  }, {
    key: "render",
    value: function render() {
      this.getCalculatedValues(this.props);
      return _react2["default"].createElement(
        "g",
        { transform: this.transform },
        this.renderTick(this.props),
        this.renderLabel(this.props)
      );
    }
  }], [{
    key: "role",
    value: "tick",
    enumerable: true
  }, {
    key: "propTypes",
    value: {
      position: _react.PropTypes.number,
      tick: _react.PropTypes.any,
      orientation: _react.PropTypes.oneOf(["top", "bottom", "left", "right"]),
      style: _react.PropTypes.object,
      label: _react.PropTypes.any
    },
    enumerable: true
  }]);

  var _Tick = Tick;
  Tick = (0, _radium2["default"])(Tick) || Tick;
  return Tick;
})(_react2["default"].Component);

exports["default"] = Tick;
module.exports = exports["default"];
