// victory-pie/lib/components/slice.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _lodash = require("lodash");

var _lodash2 = _interopRequireDefault(_lodash);

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _radium = require("radium");

var _radium2 = _interopRequireDefault(_radium);

var Slice = (function (_React$Component) {
  _inherits(Slice, _React$Component);

  function Slice() {
    _classCallCheck(this, _Slice);

    _get(Object.getPrototypeOf(_Slice.prototype), "constructor", this).apply(this, arguments);
  }

  _createClass(Slice, [{
    key: "evaluateStyle",
    value: function evaluateStyle(style) {
      var _this = this;

      return _lodash2["default"].transform(style, function (result, value, key) {
        result[key] = _lodash2["default"].isFunction(value) ? value.call(_this, _this.props.slice.data) : value;
      });
    }
  }, {
    key: "getStyles",
    value: function getStyles() {
      var dataStyles = _lodash2["default"].omit(this.props.slice.data, ["x", "y", "label"]);
      return this.evaluateStyle(_lodash2["default"].merge({}, this.props.style, dataStyles));
    }
  }, {
    key: "renderSlice",
    value: function renderSlice(props) {
      return _react2["default"].createElement("path", {
        d: props.pathFunction.call(this, props.slice),
        style: this.getStyles()
      });
    }
  }, {
    key: "render",
    value: function render() {
      return this.renderSlice(this.props);
    }
  }], [{
    key: "propTypes",
    value: {
      slice: _react.PropTypes.object,
      pathFunction: _react.PropTypes.func,
      style: _react.PropTypes.object
    },
    enumerable: true
  }]);

  var _Slice = Slice;
  Slice = (0, _radium2["default"])(Slice) || Slice;
  return Slice;
})(_react2["default"].Component);

exports["default"] = Slice;
module.exports = exports["default"];
