// victory-pie/lib/components/slice-label.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _lodash = require("lodash");

var _lodash2 = _interopRequireDefault(_lodash);

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _radium = require("radium");

var _radium2 = _interopRequireDefault(_radium);

var _victoryLabel = require("victory-label");

var SliceLabel = (function (_React$Component) {
  _inherits(SliceLabel, _React$Component);

  function SliceLabel() {
    _classCallCheck(this, _SliceLabel);

    _get(Object.getPrototypeOf(_SliceLabel.prototype), "constructor", this).apply(this, arguments);
  }

  _createClass(SliceLabel, [{
    key: "getCalculatedValues",
    value: function getCalculatedValues(props) {
      var position = props.positionFunction.call(this, props.slice);
      this.x = position[0];
      this.y = position[1];
      this.data = props.slice.data;
      this.label = this.data.label ? "" + this.evaluateProp(this.data.label) : "" + this.data.x;
    }
  }, {
    key: "evaluateStyle",
    value: function evaluateStyle(style) {
      var _this = this;

      return _lodash2["default"].transform(style, function (result, value, key) {
        result[key] = _this.evaluateProp(value);
      });
    }
  }, {
    key: "evaluateProp",
    value: function evaluateProp(prop) {
      return _lodash2["default"].isFunction(prop) ? prop.call(this, this.data) : prop;
    }
  }, {
    key: "renderLabelComponent",
    value: function renderLabelComponent(props) {
      var component = props.labelComponent;
      var style = this.evaluateStyle(_lodash2["default"].merge({ padding: 0 }, props.style, component.props.style));
      var children = component.props.children || this.label;
      var newProps = {
        x: component.props.x || this.x,
        y: component.props.y || this.y,
        data: this.data, // Pass data for custom label component to access
        textAnchor: component.props.textAnchor || "start",
        verticalAnchor: component.props.verticalAnchor || "middle",
        style: style
      };
      return _react2["default"].cloneElement(component, newProps, children);
    }
  }, {
    key: "renderVictoryLabel",
    value: function renderVictoryLabel(props) {
      var style = this.evaluateStyle(_lodash2["default"].merge({ padding: 0 }, props.style));
      return _react2["default"].createElement(
        _victoryLabel.VictoryLabel,
        {
          x: this.x,
          y: this.y,
          data: this.data,
          style: style
        },
        this.label
      );
    }
  }, {
    key: "renderLabel",
    value: function renderLabel(props) {
      return props.labelComponent ? this.renderLabelComponent(props) : this.renderVictoryLabel(props);
    }
  }, {
    key: "render",
    value: function render() {
      this.getCalculatedValues(this.props);
      return this.renderLabel(this.props);
    }
  }], [{
    key: "propTypes",
    value: {
      labelComponent: _react.PropTypes.any,
      positionFunction: _react.PropTypes.func,
      slice: _react.PropTypes.object,
      style: _react.PropTypes.object
    },
    enumerable: true
  }]);

  var _SliceLabel = SliceLabel;
  SliceLabel = (0, _radium2["default"])(SliceLabel) || SliceLabel;
  return SliceLabel;
})(_react2["default"].Component);

exports["default"] = SliceLabel;
module.exports = exports["default"];
