// victory-line/lib/components/line-segment.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _lodash = require("lodash");

var _lodash2 = _interopRequireDefault(_lodash);

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _radium = require("radium");

var _radium2 = _interopRequireDefault(_radium);

var _d3Shape = require("d3-shape");

var _d3Shape2 = _interopRequireDefault(_d3Shape);

var LineSegment = (function (_React$Component) {
  _inherits(LineSegment, _React$Component);

  function LineSegment() {
    _classCallCheck(this, _LineSegment);

    _get(Object.getPrototypeOf(_LineSegment.prototype), "constructor", this).apply(this, arguments);
  }

  _createClass(LineSegment, [{
    key: "evaluateStyle",
    value: function evaluateStyle(style) {
      var _this = this;

      return _lodash2["default"].transform(style, function (result, value, key) {
        result[key] = _this.evaluateProp(value);
      });
    }
  }, {
    key: "evaluateProp",
    value: function evaluateProp(prop) {
      return _lodash2["default"].isFunction(prop) ? prop.call(this, this.props.data) : prop;
    }
  }, {
    key: "getCalculatedValues",
    value: function getCalculatedValues(props) {
      this.style = this.evaluateStyle(props.style);
      this.interpolation = this.evaluateProp(props.interpolation);
      var xScale = props.scale.x;
      var yScale = props.scale.y;
      var lineFunction = _d3Shape2["default"].line().curve(_d3Shape2["default"][this.interpolation]).x(function (data) {
        return xScale(data.x);
      }).y(function (data) {
        return yScale(data.y);
      });
      this.path = lineFunction(props.data);
    }
  }, {
    key: "renderLine",
    value: function renderLine() {
      return _react2["default"].createElement("path", { style: this.style, d: this.path });
    }
  }, {
    key: "render",
    value: function render() {
      this.getCalculatedValues(this.props);
      return this.renderLine();
    }
  }], [{
    key: "propTypes",
    value: {
      data: _react.PropTypes.array,
      interpolation: _react.PropTypes.string,
      scale: _react.PropTypes.object,
      style: _react.PropTypes.object
    },
    enumerable: true
  }]);

  var _LineSegment = LineSegment;
  LineSegment = (0, _radium2["default"])(LineSegment) || LineSegment;
  return LineSegment;
})(_react2["default"].Component);

exports["default"] = LineSegment;
module.exports = exports["default"];
