// victory-line/lib/components/line-label.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _lodash = require("lodash");

var _lodash2 = _interopRequireDefault(_lodash);

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _radium = require("radium");

var _radium2 = _interopRequireDefault(_radium);

var _victoryLabel = require("victory-label");

var LineLabel = (function (_React$Component) {
  _inherits(LineLabel, _React$Component);

  function LineLabel() {
    _classCallCheck(this, _LineLabel);

    _get(Object.getPrototypeOf(_LineLabel.prototype), "constructor", this).apply(this, arguments);
  }

  _createClass(LineLabel, [{
    key: "evaluateStyle",
    value: function evaluateStyle(style) {
      var _this = this;

      return _lodash2["default"].transform(style, function (result, value, key) {
        result[key] = _lodash2["default"].isFunction(value) ? value.call(_this, _this.props.data) : value;
      });
    }
  }, {
    key: "renderLabelComponent",
    value: function renderLabelComponent(props) {
      var component = props.label;
      var style = this.evaluateStyle(_lodash2["default"].merge({ padding: 0 }, props.style, component.props.style));
      var children = component.props.children || "";
      var newProps = {
        x: component.props.x || props.position.x + style.padding,
        y: component.props.y || props.position.y - style.padding,
        data: props.data, // Pass data for custom label component to access
        textAnchor: component.props.textAnchor || "start",
        verticalAnchor: component.props.verticalAnchor || "middle",
        style: style
      };
      return _react2["default"].cloneElement(component, newProps, children);
    }
  }, {
    key: "renderVictoryLabel",
    value: function renderVictoryLabel(props) {
      var style = this.evaluateStyle(_lodash2["default"].merge({ padding: 0 }, props.style));
      return _react2["default"].createElement(
        _victoryLabel.VictoryLabel,
        {
          x: props.position.x + style.padding,
          y: props.position.y - style.padding,
          data: props.data,
          textAnchor: "start",
          verticalAnchor: "middle",
          style: style
        },
        props.label
      );
    }
  }, {
    key: "renderLabel",
    value: function renderLabel(props) {
      return props.label && props.label.props ? this.renderLabelComponent(props) : this.renderVictoryLabel(props);
    }
  }, {
    key: "render",
    value: function render() {
      return _react2["default"].createElement(
        "g",
        null,
        this.renderLabel(this.props)
      );
    }
  }], [{
    key: "propTypes",
    value: {
      data: _react.PropTypes.array,
      labelComponent: _react.PropTypes.any,
      position: _react.PropTypes.object,
      style: _react.PropTypes.object
    },
    enumerable: true
  }]);

  var _LineLabel = LineLabel;
  LineLabel = (0, _radium2["default"])(LineLabel) || LineLabel;
  return LineLabel;
})(_react2["default"].Component);

exports["default"] = LineLabel;
module.exports = exports["default"];
