// victory-axis/lib/components/grid.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _lodash = require("lodash");

var _lodash2 = _interopRequireDefault(_lodash);

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _radium = require("radium");

var _radium2 = _interopRequireDefault(_radium);

var GridLine = (function (_React$Component) {
  _inherits(GridLine, _React$Component);

  function GridLine() {
    _classCallCheck(this, _GridLine);

    _get(Object.getPrototypeOf(_GridLine.prototype), "constructor", this).apply(this, arguments);
  }

  _createClass(GridLine, [{
    key: "evaluateStyle",
    value: function evaluateStyle(style) {
      var _this = this;

      return _lodash2["default"].transform(style, function (result, value, key) {
        result[key] = _lodash2["default"].isFunction(value) ? value.call(_this, _this.props.tick) : value;
      });
    }
  }, {
    key: "render",
    value: function render() {
      return _react2["default"].createElement(
        "g",
        { transform: "translate(" + this.props.xTransform + ", " + this.props.yTransform + ")" },
        _react2["default"].createElement("line", {
          x2: this.props.x2,
          y2: this.props.y2,
          style: this.evaluateStyle(this.props.style)
        })
      );
    }
  }], [{
    key: "role",
    value: "grid",
    enumerable: true
  }, {
    key: "propTypes",
    value: {
      tick: _react.PropTypes.any,
      x2: _react.PropTypes.number,
      y2: _react.PropTypes.number,
      xTransform: _react.PropTypes.number,
      yTransform: _react.PropTypes.number,
      style: _react.PropTypes.object
    },
    enumerable: true
  }]);

  var _GridLine = GridLine;
  GridLine = (0, _radium2["default"])(GridLine) || GridLine;
  return GridLine;
})(_react2["default"].Component);

exports["default"] = GridLine;
module.exports = exports["default"];
