// victory-scatter/lib/components/point.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _lodash = require("lodash");

var _lodash2 = _interopRequireDefault(_lodash);

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _radium = require("radium");

var _radium2 = _interopRequireDefault(_radium);

var _pathHelpers = require("../path-helpers");

var _pathHelpers2 = _interopRequireDefault(_pathHelpers);

var _victoryLabel = require("victory-label");

var Point = (function (_React$Component) {
  _inherits(Point, _React$Component);

  function Point() {
    _classCallCheck(this, _Point);

    _get(Object.getPrototypeOf(_Point.prototype), "constructor", this).apply(this, arguments);
  }

  _createClass(Point, [{
    key: "getCalculatedValues",
    value: function getCalculatedValues(props) {
      this.style = this.getStyle(props);
    }
  }, {
    key: "getStyle",
    value: function getStyle(props) {
      var stylesFromData = _lodash2["default"].omit(props.data, ["x", "y", "z", "size", "symbol", "name", "label"]);
      var dataStyle = this.evaluateStyle(_lodash2["default"].merge({}, props.style.data, stylesFromData));
      // match certain label styles to data if styles are not given
      var matchedStyle = _lodash2["default"].pick(dataStyle, ["opacity", "fill"]);
      var padding = props.style.labels.padding || props.size * 0.25;
      var labelStyle = this.evaluateStyle(_lodash2["default"].merge({ padding: padding }, matchedStyle, props.style.labels));
      return { data: dataStyle, labels: labelStyle };
    }
  }, {
    key: "evaluateStyle",
    value: function evaluateStyle(style) {
      var _this = this;

      return _lodash2["default"].transform(style, function (result, value, key) {
        result[key] = _this.evaluateProp(value);
      });
    }
  }, {
    key: "evaluateProp",
    value: function evaluateProp(prop) {
      return _lodash2["default"].isFunction(prop) ? prop.call(this, this.props.data) : prop;
    }
  }, {
    key: "getPath",
    value: function getPath(props) {
      var pathFunctions = {
        circle: _pathHelpers2["default"].circle,
        square: _pathHelpers2["default"].square,
        diamond: _pathHelpers2["default"].diamond,
        triangleDown: _pathHelpers2["default"].triangleDown,
        triangleUp: _pathHelpers2["default"].triangleUp,
        plus: _pathHelpers2["default"].plus,
        star: _pathHelpers2["default"].star
      };
      var size = this.evaluateProp(props.size);
      var symbol = this.evaluateProp(props.symbol);
      return pathFunctions[symbol].call(this, props.x, props.y, size);
    }
  }, {
    key: "renderPoint",
    value: function renderPoint(props) {
      return _react2["default"].createElement("path", {
        style: this.style.data,
        d: this.getPath(props),
        shapeRendering: "optimizeSpeed"
      });
    }
  }, {
    key: "renderLabel",
    value: function renderLabel(props) {
      if (props.showLabels === false || !props.data.label) {
        return undefined;
      }
      var component = props.labelComponent;
      var componentStyle = component && component.props.style || {};
      var style = _lodash2["default"].merge({}, this.style.labels, componentStyle);
      var children = component && component.props.children || props.data.label;
      var labelProps = {
        x: component && component.props.x || props.x,
        y: component && component.props.y || props.y - style.padding,
        dy: component && component.props.dy,
        data: props.data,
        textAnchor: component && component.props.textAnchor || style.textAnchor,
        verticalAnchor: component && component.props.verticalAnchor || "end",
        style: style
      };

      return component ? _react2["default"].cloneElement(component, labelProps, children) : _react2["default"].createElement(_victoryLabel.VictoryLabel, labelProps, children);
    }
  }, {
    key: "render",
    value: function render() {
      this.getCalculatedValues(this.props);
      return _react2["default"].createElement(
        "g",
        null,
        this.renderPoint(this.props),
        this.renderLabel(this.props)
      );
    }
  }], [{
    key: "propTypes",
    value: {
      data: _react.PropTypes.shape({
        x: _react2["default"].PropTypes.any,
        y: _react2["default"].PropTypes.any
      }),
      labelComponent: _react2["default"].PropTypes.element,
      symbol: _react.PropTypes.oneOfType([_react.PropTypes.oneOf(["circle", "diamond", "plus", "square", "star", "triangleDown", "triangleUp"]), _react.PropTypes.func]),
      size: _react.PropTypes.oneOfType([_react.PropTypes.number, _react.PropTypes.func]),
      showLabels: _react2["default"].PropTypes.bool,
      style: _react.PropTypes.shape({
        data: _react2["default"].PropTypes.object,
        labels: _react2["default"].PropTypes.object
      }),
      x: _react2["default"].PropTypes.number,
      y: _react2["default"].PropTypes.number
    },
    enumerable: true
  }, {
    key: "defaultProps",
    value: {
      showLabels: true
    },
    enumerable: true
  }]);

  var _Point = Point;
  Point = (0, _radium2["default"])(Point) || Point;
  return Point;
})(_react2["default"].Component);

exports["default"] = Point;
module.exports = exports["default"];
